### PR TITLE
feat(ace): replace per-order ACE circuit registries with one order-invariant circuit per relation

### PR DESCRIPTION
**Pull request by Nashtare**
## Summary

  Replaces the per-proof-order ACE circuit registries (a Merkle tree of one circuit per AIR ordering, indexed by Lehmer rank) with **one canonical, order-invariant ACE circuit per relation**, for both the Miden VM and the
  precompiles VM. Proof order is now carried by *where* the recursive verifier scatters its out-of-domain rows and auxiliary-bus boundary values at ingest, not by *which* registry leaf it opens.

  Both relation digests are re-minted over the circuit digest instead of a registry root. **Recursive proofs produced by earlier releases no longer verify and must be regenerated.**

  ## Motivation

  The registry model required generating, committing, and authenticating `N!` circuit variants (24 for the VM, a 10!-sized leaf space for the PVM), plus a ~4.1k-line generated leaf table, plus a Lehmer-tag routing surface
  inside the verifier that was soundness-adjacent and hard to review. All of that existed only because the ACE circuit's READ layout was laid out in proof order while the proof commits traces in height-sorted order. Moving
  the reordering to ingest time makes the circuit itself a fixed protocol constant.

  ## Design

  - **Canonical circuit.** `build_canonical_multi_air_ace_circuit` places every AIR's trace regions at its instance-order offset and gives each AIR a dedicated fold-coefficient READ slot (`InputKey::MultiAirFoldCoeff(k)`),
  so the same circuit serves every proof order. One circuit, one digest: `ACE_CIRCUIT_DIGEST` is both the advice-map key and the value the MASM loader pins the stream to.
  - **Fold coefficients staged in MASM.** The native verifier folds constraint roots as a Horner chain over proof order, so proof position `j` carries `β^(N-1-j)`. `stage_air_fold_coefficients` computes `pos(k)` in-VM from
  the committed heights via the shared `proof_order_position_from_heights` primitive and stages `β^(N-1-pos(k))` into AIR `k`'s slot. No exponent is baked into the circuit.
  - **OOD row scatter.** The transcript absorb and DEEP Horner accumulation stay positional over the advice stream; only the `adv_pipe` destination is retargeted at segment boundaries, through a per-position dispatch table
  and `dynexec` onto fixed-length `pipe_k` procedures (segment lengths vary per AIR, and `repeat.N` needs a compile-time immediate).
  - **σ (aux-bus boundary) scatter.** Boundary values are gathered from their proof-order slots and stored at canonical addresses with pure arithmetic addressing; no dispatch needed since every value is one word. The PVM
  handles its two composite AIRs (`EidosCompression`, `BytePairAnd8`), which expose two boundary values each.
  - **Regeneration.** Both relations build the shipped circuit through the same function the regen tool uses; drift is caught by `make check-constraints` / `make check-pvm-constants`, and the compiled-in digests are pinned
  to a fresh build by unit tests.

  ## Commits

  1. `feat(ace-codegen): add canonical multi-AIR ACE circuit builder` — layout, pipeline, air-crate wrapper, fold-coefficient staging, and the restored direct tests for `proof_order_position_from_heights`.
  2. `feat(vm): evaluate the canonical ACE circuit in the recursive verifier` — VM OOD/σ scatters, single-segment circuit authentication, relation digest re-mint, registry deletion.
  3. `feat(pvm): evaluate the canonical ACE circuit in the recursive verifier` — PVM mirror (22 segments, composite AIRs), `pvm_ood_frames.rs` renderer, deletion of the factored/registry machinery, tooling renames.
  4. `docs(ace): describe the order-invariant cutover and drop stale registry references`
  5. `chore(ace): regenerate VM relation/circuit digests after eidos rebase`

  ## Breaking changes

  - Removed the registry and factored-circuit APIs from `miden_ace_codegen` and `miden_air::ace`, plus `miden_air::config::{ACE_CIRCUIT_REGISTRY_ROOT, ACE_CIRCUIT_REGISTRY_DEPTH, ACE_CIRCUIT_REGISTRY_LEAF_COUNT,
  ace_registry_path}` and `miden_air::PROOF_ORDER_REGISTRY_DEPTH`.
  - `miden_air::config::ACE_CIRCUIT_DIGEST` replaces the registry root; `relation_digest`'s second argument is now that circuit digest. `RecursiveAceCircuit` drops its `shuffle_prefix_len` / `shuffle_commitment` /
  `common_commitment` fields and `commitment` now means the digest of the whole stream.
  - `MAX_REGISTRY_AIRS` → `MAX_ORDER_AIRS`; the Lehmer helpers move to a `proof_order` module (still re-exported).
  - `miden-precompiles-prover`: feature `registry-tools` → `constants-tools`, binary `pvm-registry-regen` → `pvm-constants-regen`, module `ace_registry_regen` → `ace_constants_regen`. Make targets are now
  `regenerate-pvm-constants` / `check-pvm-constants`.
  - Both `RELATION_DIGEST` constants (VM and PVM) change value.

  ## Performance

  Measured end-to-end on `verify_vm_proof` (4 AIRs) against the registry-based base, same synthetic inner proof:

  | Inner proof | base (registry) | this PR (canonical + scatter) | delta |
  |---|---|---|---|
  | small | 44,367 cycles | 46,306 cycles | +1,939 (+4.4%) |
  | large | 44,305 cycles | 46,244 cycles | +1,939 (+4.4%) |

  The delta is identical across proof sizes: it is a fixed cost driven by AIR/segment count, not trace height, so it shrinks as a fraction on production-size proofs. It comes from the per-segment `dynexec` dispatch in the
  OOD scatter (isolated microbenchmark, one row):

  | | flat, no scatter | `dynexec` dispatch (shipped) | per-block `while` loop | `if/else` cascade |
  |---|---|---|---|---|
  | VM (10 segments) | 248 | 476 (+22/segment) | 1,367 (+451%) | 568 |
  | PVM (22 segments) | 704 | 1,256 (+25/segment) | 3,765 (+435%) | n/a |

  Both alternatives were prototyped and rejected; the benchmarks are checked in as tests with hard cycle ceilings so a "simplification" back to a loop fails CI.

  ## Testing

  - Rust circuit equivalence: canonical vs. per-order builder across all 24 VM orders (roots pairwise distinct) and ~95 structured PVM orders.
  - Generated OOD hooks driven against a Rust oracle for all 24 VM orders + a tie, and 34 structured PVM orders + a tie; σ scatter 24/24 VM, 34 PVM.
  - `proof_order_position_from_heights`: 13-case partial-tie table, 64-case proptest against `sort_by_key(|i| (h[i], i))` checking every AIR's position, out-of-range-height trap.
  - Negative order binding: a mutant `verify_vm_proof` with the scatter table left unstaged rejects a non-identity-order fixture (`ProcedureNotFound` on the zeroed dispatch slot), so the mechanism is shown to be load-bearing
  independently of transcript divergence. Every advice-supplied height is forgery-tested on both relations.
  - Digest pins: compiled-in circuit digest == fresh canonical build, relation digest == `relation_digest(protocol_id, circuit_digest)`, VM/PVM protocol ids and digests pairwise distinct.
  - Full suite: 5895 passed, 30 skipped. `make check-constraints` and `make check-pvm-constants` clean. Snapshots updated.

  ## Size

  | | added | removed | net |
  |---|---|---|---|
  | Production Rust + MASM | 2,609 | 7,593 | −4,984 |
  | …of which the generated PVM registry leaf table | 0 | 4,138 | −4,138 |
  | Tests + snapshots | 3,191 | 1,238 | +1,953 |
  | Generated MASM | 1,011 | 609 | +402 |
  | **Total (83 files)** | **6,853** | **9,495** | **−2,642** |

Out of which, 3,191 lines added for testing (so raw difference: 5,833 lines removed).

  ## Merge notes

  - The removed CI job was a named check, `check PVM registry and MASM artifacts`; its step now runs inside `check-constraints`. Any branch-protection rule requiring the old check by name needs updating.
  - Possible follow-up: the two scatter benchmarks near-duplicate their harness across VM and PVM; a shared harness would cut ~600–800 lines.

### CODE REVIEW

**Comment by Al-Kindi-0** [comment]
Looks very promising and I was investigating this for some time now. I think the idea is the right one and it probably does make sense to remove the registry even if at a small cost, though I think we can actually do better

I am in the process of rebasing https://github.com/0xMiden/miden-vm/pull/3756 which I think should be the base for this PR.

I am also working on a followup/stacked PR to this one integrating some further optimizations that could improve things.

**Comment by Nashtare** [comment]
> I am in the process of rebasing https://github.com/0xMiden/miden-vm/pull/3756 which I think should be the base for this PR.

Fair point, I'll rebase this branch once updated!

**Comment by Al-Kindi-0** [comment]
@Nashtare , https://github.com/0xMiden/miden-vm/pull/3756 is rebased and ready for your rebase of this PR on it

**Comment by Nashtare** [comment]
> @Nashtare , #3756 is rebased and ready for your rebase of this PR on it

Already did!

**Review by Al-Kindi-0 — APPROVED** [review]
LGTM!

File: crates/ace-codegen/src/order_maps.rs [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+         #! `id_by_pos` (instance index at each proof position) and `pos_by_id` (proof position of \
+         each\n\
+         #! instance). Equal heights order by instance index, exactly like the native \
+         `TraceOrder`.\n\
+         #!\n\
+         #! Must run after every height has been bounded; keys are compared as u32 values.\n\
+         #!\n\
+         #! Inputs:  []\n\
+         #! Outputs: []\n\
+         pub proc stage_proof_order_maps\n\
```
Comment (huitseeker) [inline crates/ace-codegen/src/order_maps.rs]: nit: `render_proof_order_maps` emits a public `stage_proof_order_maps` procedure without the required `#! Invocation: exec` line.

Comment (Al-Kindi-0) [reply crates/ace-codegen/src/order_maps.rs]: fixed

File: crates/ace-codegen/src/order_maps.rs [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+         #! `id_by_pos` (instance index at each proof position) and `pos_by_id` (proof position of \
+         each\n\
+         #! instance). Equal heights order by instance index, exactly like the native \
+         `TraceOrder`.\n\
+         #!\n\
+         #! Must run after every height has been bounded; keys are compared as u32 values.\n\
+         #!\n\
+         #! Inputs:  []\n\
+         #! Outputs: []\n\
+         pub proc stage_proof_order_maps()\n\
```
Comment (huitseeker) [inline crates/ace-codegen/src/order_maps.rs:417]: nit: The generated public procedure still lacks the required `#! Invocation: exec` line, although the earlier comment was marked fixed.

File: crates/precompiles-verifier/src/pvm_ood_frames.rs [hunk 3]
```diff
@@ -0,0 +1,728 @@
...
+                "the PVM out-of-domain row is {} felts, which is not {ADV_PIPE_BLOCK_FELTS}-felt \
+                 aligned",
+                geometry.row_felts()
+            ));
+        }
+
+        Ok(geometry)
+    }
+
+    fn groups(&self) -> [(&'static str, &Vec<usize>); 3] {
```
Comment (huitseeker) [inline crates/precompiles-verifier/src/pvm_ood_frames.rs]: nit: could change from `&Vec<usize>` to `&[usize]`

File: air/src/ace/multi_air.rs [hunk 4]
```diff
@@ -10,54 +10,62 @@ const LMCS_ALIGNMENT: usize = 8;
...
+/// verifier evaluates the order-invariant circuit instead; this construction is the independent
+/// reference that one is tested against.
 pub fn build_multi_air_ace_circuit_for_order(
     config: AceConfig,
     order: &ProofOrder,
 ) -> Result<AceCircuit<QuadFelt>, AceError> {
-    build_factored_multi_air_ace_circuit(config)?.circuit_for_order(order)
+    let airs = AIRS.map(HandwrittenMidenAir);
+    let proof_order: Vec<usize> = order.airs().iter().map(|air| air.instance_index()).collect();
+    miden_ace_codegen::build_multi_air_ace_circuit(&airs, &proof_order, config, LMCS_ALIGNMENT)
```
Comment (adr1anh) [inline air/src/ace/multi_air.rs:23]: This is only called by tests now. Could we replace those uses and delete both this wrapper and `build_multi_air_ace_circuit` in a follow-up?

The existing `eval_folded_constraints` helper might let us compare against a Horner fold of the AIR results minus `q * v`, keeping the same proof orders and nonzero quotient inputs. That still needs each AIR's selectors and periodic values. `multi_air_eval_circuit_masm` could use the canonical builder.

File: crates/ace-codegen/src/layout/policy.rs [hunk 5]
```diff
@@ -27,11 +27,14 @@ struct LayoutPolicy {
...
 
-/// Whether the layout includes the slots needed to combine multiple AIR instances.
+/// Whether the layout includes the slots needed to combine multiple AIR instances, and whether
+/// fold coefficients get a real READ slot (`MultiCanonical`) or are baked into the circuit's
+/// gates for one fixed proof order (`Multi`, the per-order oracle composition).
 #[derive(Clone, Copy)]
 enum AirComposition {
     Single,
     Multi { air_count: NonZeroUsize },
+    MultiCanonical { air_count: NonZeroUsize },
```
Comment (adr1anh) [inline crates/ace-codegen/src/layout/policy.rs:37]: If we remove the old builder, can we also drop the single-AIR builder's `num_airs >= 2` mode? Then `Multi` and `MultiCanonical` could become one variant, and `MultiAirIndices::fold_coeff_start` would no longer need to be an `Option`.

File: crates/ace-codegen/src/pipeline.rs [hunk 6]
```diff
@@ -370,122 +375,30 @@ where
...
     let num_airs = airs.len();
-    if num_airs == 0 || config.num_airs != num_airs {
+    if num_airs == 0 || num_airs > MAX_ORDER_AIRS || config.num_airs != num_airs {
         return Err(AceError::InvalidInputLayout {
             message: format!(
-                "multi-AIR composition requires a nonempty airs slice and matching num_airs; got \
-                 {} AIRs and num_airs {}",
+                "multi-AIR composition requires a nonempty airs slice of at most \
+                 {MAX_ORDER_AIRS} AIRs and matching num_airs; got {} AIRs and num_airs {}",
                 num_airs, config.num_airs
```
Comment (adr1anh) [inline crates/ace-codegen/src/pipeline.rs:402]: Can we derive `num_airs` from `airs.len()` and remove it from `AceConfig`? This also means removing the single-AIR builder's option to use a multi-AIR layout.

Once the old builder is gone, we can drop `canonical_order` and the `order` argument to `accumulate_block_offsets`. The `MAX_ORDER_AIRS` check here also seems worth revisiting: this builder no longer ranks permutations. The MASM order-map limits would still need to be checked.

File: crates/lib/core/asm/stark/constraints_eval_inputs.masm [hunk 7]
```diff
@@ -20,9 +20,17 @@ const SELECTOR_STRIDE = 6
 #!   Word 4 (slots 8-9): f, s0
 #!
 #! The multi-AIR layout then appends:
-#!   slot 10: multi-air fold beta (sampled by the generic random coin)
+#!   slot 10: multi-air fold beta (sampled by the generic random coin). Staged for every relation,
+#!            but the canonical circuit folds with the per-AIR MultiAirFoldCoeff slots below
+#!            instead of this challenge directly; it stays a READ slot because AIR k's selector
+#!            triple is pinned at FIRST_SELECTOR_OFFSET + SELECTOR_STRIDE * k, and removing it
+#!            would shift every selector offset and re-mint both relation digests.
```
Comment (adr1anh) [inline crates/lib/core/asm/stark/constraints_eval_inputs.masm:27]: Can we remove this slot? `stage_air_fold_coefficients` already reads beta directly, and the canonical circuit uses `MultiAirFoldCoeff`.

We'd need to update the offsets and regenerate both relation digests, so it makes sense to group this with the other layout changes. The coefficient test in `ace_read_check.rs` could read beta from `COMPOSITION_COEF_PTR + 2` instead.

File: crates/lib/core/asm/stark/constraints_eval_inputs.masm [hunk 8]
```diff
@@ -20,9 +20,17 @@ const SELECTOR_STRIDE = 6
...
 #!   slots 11 + 3k .. 13 + 3k: AIR k's lifted selectors, canonical instance order
```
Comment (adr1anh) [inline crates/lib/core/asm/stark/constraints_eval_inputs.masm:28]: Do we still need to compute and store the global `IsFirst`, `IsLast` and `IsTransition` values here? The canonical circuit uses the per-AIR selectors.

We'd have to keep the `z_k` store in the same block, and keep the global selectors available for single-AIR circuits.

File: crates/ace-codegen/src/masm.rs [hunk 9]
```diff
@@ -24,28 +24,95 @@ pub struct MasmConstraintsEvalConfig<'a> {
...
-    /// Number of valid proof-order tags; higher registry slots are padding.
-    pub order_tag_count: usize,
     /// Number of AIR instances in the relation.
     pub num_airs: usize,
+    /// How the evaluator stages one multi-AIR fold coefficient per AIR, or `None` for a relation
+    /// whose ACE READ layout reserves no coefficient slots.
+    ///
+    /// The staged block sits immediately after the selectors, so a relation without those slots
+    /// would write past its `auxiliary_ace_inputs_ptr` region.
+    pub fold_coefficients: Option<FoldCoefficientStaging<'a>>,
```
Comment (adr1anh) [inline crates/ace-codegen/src/masm.rs:36]: Both VM and PVM set `fold_coefficients` to `Some`. Is there another caller that needs `None`? Otherwise we could make this required and remove that branch and its tests.

File: crates/precompiles-verifier/src/ace.rs [hunk 10]
```diff
@@ -72,53 +64,90 @@ pub fn build_precompile_multi_air_ace_circuit() -> Result<AceCircuit<QuadFelt>,
...
+    pub num_inputs: usize,
+    /// Number of ACE EVAL rows.
+    pub num_eval_gates: usize,
+    /// Encoded instruction stream length in base-field elements.
+    pub stream_len: usize,
+    /// Eidos digest of the full instruction stream: the advice-map key and the value pinned by
+    /// the compiled-in circuit digest.
+    pub commitment: Word,
+    /// Encoded ACE instruction stream consumed by `eval_circuit`.
+    pub instructions: Vec<Felt>,
```
Comment (adr1anh) [inline crates/precompiles-verifier/src/ace.rs:97]: `PvmRecursiveAceCircuit` and `RecursiveAceCircuit` have the same fields, and both builders copy the same data out of `EncodedCircuit`. Could they share a type, or just the conversion? The VM/PVM configs and caches can stay separate.

File: crates/precompiles-verifier/src/ace.rs [hunk 11]
```diff
@@ -72,53 +64,90 @@ pub fn build_precompile_multi_air_ace_circuit() -> Result<AceCircuit<QuadFelt>,
...
+    let instructions = encoded.instructions();
+    let stream_len = encoded.size_in_felt();
+    if stream_len != instructions.len() {
+        return Err(AceError::InvalidInputLayout {
+            message: format!(
+                "ACE circuit stream length ({stream_len}) does not match instruction count ({})",
+                instructions.len()
+            ),
+        });
+    }
```
Comment (adr1anh) [inline crates/precompiles-verifier/src/ace.rs:115]: `size_in_felt()` already returns `instructions.len()`, so this check can never fail. We can delete it here and in `air/src/ace/recursive.rs`. The 8-felt alignment check below is still useful.

File: crates/lib/core/tests/stark/vm_scatter_bench.rs [hunk 12]
```diff
@@ -0,0 +1,854 @@
...
+    program("", &setup, &measured, row_felts)
+}
+
+/// The deployed mechanism, exactly as `constraints_regen.rs` renders it.
+///
+/// One `pipe_k` procedure per distinct segment length, dispatched per stream position through a
+/// `(row-relative destination, digest address)` table. Groups holding at most one non-empty
+/// segment carry no proof-order freedom, so they are emitted with a compile-time destination and
+/// a direct `exec` instead of a table entry.
+fn production_source(geometry: &RowGeometry, order: &[usize], row_felts: u32) -> String {
```
Comment (adr1anh) [inline crates/lib/core/tests/stark/vm_scatter_bench.rs:356]: Could we remove `production_source`, `while_source`, and their comparison tests here and in `pvm_scatter_bench.rs`? They generate their own MASM, so they can pass even if `ood_frames.masm` changes.

`generated_hook_scatters_both_rows_to_canonical_addresses` runs the generated code and checks memory, the Eidos chaining word, the DEEP accumulator, pointers and ingest cycles. I'd keep that test and move over any checks it is missing.

File: crates/lib/core/tests/stark/vm_scatter_bench.rs [hunk 13]
```diff
@@ -0,0 +1,854 @@
...
+/// The checked-in hook must dispatch per segment through `dynexec`, with one `pipe_k` procedure
+/// per distinct segment length, and must contain no loop.
+///
+/// The cost comparison in `dispatched_scatter_stays_far_cheaper_than_a_per_block_loop` measures
+/// test-local renderings of each mechanism, so on its own it would stay green even if the
+/// renderer emitted a per-block loop. This reads the generated file itself.
+#[test]
+fn generated_hook_dispatches_per_segment_rather_than_looping_per_block() {
+    let geometry = RowGeometry::vm();
+    let plan = geometry.scatter_plan(&(0..MIDEN_AIR_COUNT).collect::<Vec<_>>());
```
Comment (adr1anh) [inline crates/lib/core/tests/stark/vm_scatter_bench.rs:817]: Could we keep the checks on results and cycle counts, and drop the exact `dynexec`/`exec.pipe_*` counts, helper names and ban on `while`?

The current cycle limit only covers OOD ingest. If we also want to protect staging and sigma-scatter costs, we'd need separate limits; those aren't covered today. The sigma tests' checks on memory regions and scattered values should stay.

File: crates/lib/core/tests/stark/vm_sigma_scatter.rs [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+/// This is the premise the scatter is defined against; without it the permutation asserted below
+/// would be measured from the wrong baseline.
+#[test]
+fn the_absorb_alone_stores_the_boundary_values_in_wire_order() {
+    for order in permutations(MIDEN_AIR_COUNT) {
+        let heights = heights_for_order(&order);
+        assert_eq!(
+            run(&heights, false),
+            wire_values(),
+            "the control absorb reordered the wire for heights {heights:?}"
```
Comment (adr1anh) [inline crates/lib/core/tests/stark/vm_sigma_scatter.rs:196]: `scatter = false` skips `scatter_aux_bus_boundary` and tests the loads/stores written in this file, plus `stage_proof_order_maps`. The `true` cases already exercise those maps and check the scattered values.

Could we remove this test, the equivalent PVM test, and the `scatter` flag? Keep the stack checks and proof-order cases in the remaining tests.

File: crates/lib/core/tests/stark/vm_sigma_scatter.rs [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+/// value while slot `k` carried AIR `k`'s height — a silent mispairing for every non-identity
+/// order, which is what this pins.
+#[test]
+fn the_weighted_sum_pairs_each_height_with_its_own_airs_boundary_value() {
+    let wire = wire_values();
+    for order in permutations(MIDEN_AIR_COUNT) {
+        let heights = heights_for_order(&order);
+        // The production absorb-and-scatter, then the very push the weighted sum consumes,
+        // captured in memory instead of folded into a sum.
+        let body = source(&heights, true);
```
Comment (adr1anh) [inline crates/lib/core/tests/stark/vm_sigma_scatter.rs:254]: Can we add the `push_canonical_log_heights` checks to `the_scatter_moves_every_boundary_value_to_its_canonical_air`? That would check heights and boundary values in one run per proof order and let us delete this second loop.

Both tests assemble the sequence themselves; this one doesn't test the call order inside `observe_aux_trace`.

File: crates/precompiles-verifier/src/ace.rs [hunk 16]
```diff
@@ -811,4 +978,104 @@ mod tests {
...
+            })
+            .collect();
+        let aux_refs: Vec<&[QuadFelt]> = aux_values.iter().map(Vec::as_slice).collect();
+
+        let assertions = multi_air
+            .eval_external(&challenges, &[Felt::ZERO; 4], &[], &aux_refs, &[0; NUM_CHIPLETS])
+            .expect("fixed boundary denominators are non-zero for the fixture");
+
+        assert_eq!(assertions.len(), 1, "the relation exposes exactly one external assertion");
+        assert_ne!(assertions[0], QuadFelt::ZERO, "the closure fixture must be non-vacuous");
```
Comment (adr1anh) [inline crates/precompiles-verifier/src/ace.rs:1017]: The next test, `external_assertion_matches_the_protocol_version`, uses the same inputs and checks `actual.as_slice() == &[expected]`. That already checks there's exactly one assertion and that it has the expected nonzero value. I think we can delete this test.

File: crates/precompiles-verifier/src/pvm_ood_frames.rs [hunk 17]
```diff
@@ -0,0 +1,728 @@
...
+/// per-chiplet, and is likewise fixed.
+fn pvm_scatter_plan(geometry: &PvmOodGeometry) -> Result<ScatterPlan, String> {
+    let mut dispatches = Vec::new();
+    let mut sources = Vec::new();
+    let mut group_base = 0usize;
+    let mut slot = 0usize;
+
+    for (group, widths) in geometry.groups() {
+        let occupied = widths.iter().filter(|width| **width > 0).count();
+        // Slots of one group are contiguous and indexed by proof position, so every source in the
```
Comment (adr1anh) [inline crates/precompiles-verifier/src/pvm_ood_frames.rs:269]: This looks very similar to `vm_scatter_plan` in `constraints_regen.rs`. The renderer below also repeats the code for `pipe_k`, the dispatch table and OOD ingest procedures. Could we share that code in a follow-up?

We'd still need to account for where VM/PVM place the proof-order maps, and keep PVM's alignment check.

File: crates/precompiles-verifier/src/ace_constants_regen.rs [hunk 18]
```diff
@@ -0,0 +1,1006 @@
...
+    }
+}
+
+/// Build the canonical circuit and derive every generated artifact from it.
+fn compute() -> Result<GeneratedArtifacts, String> {
+    let circuit = build_pvm_recursive_verifier_ace_circuit().map_err(|e| format!("{e}"))?;
+    let input_layout = crate::ace::build_canonical_precompile_ace_circuit()
+        .map_err(|e| format!("{e}"))?
+        .layout()
+        .clone();
```
Comment (adr1anh) [inline crates/precompiles-verifier/src/ace_constants_regen.rs:249]: We build the circuit twice here: once for the encoded instructions, then again for `InputLayout`. Could we keep the layout from the first build? `constraints_regen.rs` does the same through `recursive_verifier_input_layout`.

File: crates/precompiles-verifier/src/pvm_ood_frames.rs [hunk 19]
```diff
@@ -0,0 +1,402 @@
...
+        assert_eq!(geometry.aux, vec![40, 40, 24, 8, 24, 56, 8, 16, 24, 32]);
+        assert_eq!(geometry.quotient, 8);
+        assert_eq!(geometry.row_felts(), 1_600);
+        assert_eq!(geometry.row_blocks(), 200);
+
+        let plan = pvm_scatter_plan(&geometry).expect("scatter plan");
+        assert_eq!(plan.segment_count(), 22, "one segment per occupied per-chiplet block");
+        assert_eq!(plan.dispatched_slots(), 20, "main and aux are the order-dependent groups");
+        assert_eq!(
+            pvm_scatter_table_layout(&geometry).unwrap(),
```
Comment (huitseeker) [inline crates/precompiles-verifier/src/pvm_ood_frames.rs:312]: `pvm_row_geometry_is_the_one_the_hook_was_rendered_from` is compiled under `cfg(test)` but calls `pvm_scatter_table_layout`, which is absent unless `constants-tools` is enabled. This breaks `cargo check --benches --features internal`. Could the function also use `cfg(any(test, feature = "constants-tools"))`?

File: crates/ace-codegen/src/layout/mod.rs [hunk 20]
```diff
@@ -48,7 +48,8 @@
 //! - `quotient_next` is included in the READ layout and is mapped via
 //!   `InputKey::QuotientChunkCoord` with `offset = 1`.
 //! - `stark_vars` reserves 10 EF slots for the canonical verifier inputs. Multi-AIR layouts append
-//!   one beta slot plus three selector slots per AIR instance.
+//!   one beta slot plus three selector slots per AIR instance; canonical compositions further
```
Comment (huitseeker) [inline crates/ace-codegen/src/layout/mod.rs]: nit: `stark_vars` no longer adds a shared beta slot. `AirComposition::Multi` allocates three selectors plus one fold coefficient per AIR. Could this say that directly and drop the stale beta-slot claim?

**Review by huitseeker — COMMENTED** [review]
Overall the Rust LGTM, would love to see a review by @Al-Kindi-0

File: crates/ace-codegen/src/order_maps.rs (hunk 2 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/order_maps.rs:417]: Added

File: crates/precompiles-verifier/src/pvm_ood_frames.rs (hunk 19 above)
Comment (Al-Kindi-0) [reply crates/precompiles-verifier/src/pvm_ood_frames.rs:312]: Changed the gate to `#[cfg(any(test, feature = "constants-tools"))]`.

File: air/src/ace/multi_air.rs (hunk 4 above)
Comment (Al-Kindi-0) [reply air/src/ace/multi_air.rs]: Done

File: crates/ace-codegen/src/layout/policy.rs (hunk 5 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/layout/policy.rs]: Done

File: crates/ace-codegen/src/pipeline.rs (hunk 6 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/pipeline.rs]: Done

File: crates/lib/core/asm/stark/constraints_eval_inputs.masm (hunk 7 above)
Comment (Al-Kindi-0) [reply crates/lib/core/asm/stark/constraints_eval_inputs.masm]: Done

File: crates/lib/core/asm/stark/constraints_eval_inputs.masm (hunk 8 above)
Comment (Al-Kindi-0) [reply crates/lib/core/asm/stark/constraints_eval_inputs.masm]: Done

File: crates/ace-codegen/src/masm.rs (hunk 9 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/masm.rs]: Done

File: crates/precompiles-verifier/src/ace.rs (hunk 10 above)
Comment (Al-Kindi-0) [reply crates/precompiles-verifier/src/ace.rs]: Done

File: crates/precompiles-verifier/src/ace.rs (hunk 11 above)
Comment (Al-Kindi-0) [reply crates/precompiles-verifier/src/ace.rs]: Done

File: crates/lib/core/tests/stark/vm_scatter_bench.rs (hunk 12 above)
Comment (Al-Kindi-0) [reply crates/lib/core/tests/stark/vm_scatter_bench.rs]: Done

File: crates/lib/core/tests/stark/vm_scatter_bench.rs (hunk 13 above)
Comment (Al-Kindi-0) [reply crates/lib/core/tests/stark/vm_scatter_bench.rs]: Done. The cycle limit covers processing the two OOD rows, excluding staging and sigma scatter.

File: crates/lib/core/tests/stark/vm_sigma_scatter.rs (hunk 14 above)
Comment (Al-Kindi-0) [reply crates/lib/core/tests/stark/vm_sigma_scatter.rs]: Done

File: crates/lib/core/tests/stark/vm_sigma_scatter.rs (hunk 15 above)
Comment (Al-Kindi-0) [reply crates/lib/core/tests/stark/vm_sigma_scatter.rs]: Done

File: crates/precompiles-verifier/src/ace.rs (hunk 16 above)
Comment (Al-Kindi-0) [reply crates/precompiles-verifier/src/ace.rs]: Done

File: crates/precompiles-verifier/src/pvm_ood_frames.rs (hunk 17 above)
Comment (Al-Kindi-0) [reply crates/precompiles-verifier/src/pvm_ood_frames.rs]: Done

File: crates/precompiles-verifier/src/ace_constants_regen.rs (hunk 18 above)
Comment (Al-Kindi-0) [reply crates/precompiles-verifier/src/ace_constants_regen.rs]: Done

File: crates/ace-codegen/src/layout/mod.rs (hunk 20 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/layout/mod.rs]: Done
